### PR TITLE
[Bugfix] Fix GPT script

### PR DESCRIPTION
**Pull request by szhengac**
## Description ##
This pr fixed the following issues:
1. Data broadcast is not hooked to submod_0 after pipeline partition and pipeline module creation. Also, rank 0 is not the correct src rank for broadcast, where it take as input a global rank. This pr provides a temporary fix, while a long term solution is tracked in https://github.com/awslabs/slapo/issues/24.
2. Data samples in the first and last stages are not consistent causing output-label mismatch
3. Embedding sharding post-hook cannot directly use an all-reduce collective since it does not handle the backward. I added an AutoGrad function to enable identity mapping for the backward.

## Checklist ##

- [x] PR's title starts with a category (e.g. [Bugfix], [Model], [Tutorial], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### CODE REVIEW

File: examples/gpt/deepspeed_hf.py [hunk 1]
```diff
@@ -125,26 +130,36 @@ def train(args):
...
+
+        group_src_rank = _get_global_rank(sch.group, 0)
+
+        def broadcast_input(module, _inputs):
+            for inp in _inputs:
+                dist.broadcast(inp, src=group_src_rank, group=sch.group)
+            return _inputs
+
+        if model.mpu.get_pipe_parallel_rank() == 0:
+            sch["submod_0"].sync(mode="fwd_pre", sync_op_or_fn=broadcast_input)
```
Comment (comaniac) [inline examples/gpt/deepspeed_hf.py:162]: This programming model is weird. Users should not add more schedules after `.build`. In other word, `.build` should not return schedule. Since users have used this primitive in schedule, we should make sure this hook is preserved when tracing the module instead of asking users to specify again.

File: examples/gpt/deepspeed_hf.py [hunk 2]
```diff
@@ -165,12 +180,23 @@ def loss_fn(outputs, labels):
         model = model.to(device)
     report_memory(msg="After building model")
 
-    random_seed = 2000 + dist.get_rank()
+    if args.disable_pipeline or args.sequence_parallel:
+        random_seed = 2013 + dist.get_rank()
+    else:
+        random_seed = (
+            2013
+            + 100 * model.mpu.get_pipe_parallel_rank()
```
Comment (comaniac) [inline examples/gpt/deepspeed_hf.py:188]: Where does `.mpu` come from? Is this a DeepSpeed engine API?

File: examples/albert/deepspeed_hf.py [hunk 3]
```diff
@@ -112,7 +112,7 @@ def loss_fn(outputs, labels):
             lm_loss = lm_loss.contiguous().mean()
             return lm_loss
 
-        model, _ = slapo.build(
+        (model, _), sch = slapo.build(
```
Comment (chhzh123) [inline examples/albert/deepspeed_hf.py:115]: Why do we need to return a schedule? Users can still access `sch` even we don't return it

File: examples/gpt/deepspeed_hf.py (hunk 2 above)
Comment (szhengac) [reply examples/gpt/deepspeed_hf.py:188]: yeah, this is a deepspeed api, since model is a deepspeed object after pipeline build

File: examples/albert/deepspeed_hf.py (hunk 3 above)
Comment (szhengac) [reply examples/albert/deepspeed_hf.py:115]: i think a different sch object is created in the build?

File: examples/gpt/schedule.py [hunk 4]
```diff
@@ -414,9 +416,11 @@ def order_args_fn(*args, **kwargs):
 
 
 def broadcast_input(sch):
-    def broadcast_input(inputs):
+    group_src_rank = _get_global_rank(sch.group, 0)
+
+    def _broadcast_input(module, inputs):
         for inp in inputs:
```
Comment (zarzen) [inline examples/gpt/schedule.py:422]: Some of the `inputs` are None, when `--disable_pipeline` is used, and `num_mp = 8`

File: examples/gpt/deepspeed_hf.py (hunk 1 above)
Comment (comaniac) [reply examples/gpt/deepspeed_hf.py:162]: Please check out #27 and see if that meets the requirements of this PR. If so, please rebase this PR after merging that.

File: examples/gpt/schedule.py (hunk 4 above)
Comment (szhengac) [reply examples/gpt/schedule.py:422]: we will leave this one to the pr from you.

**Review by comaniac — APPROVED** [review]
Otherwise LGTM.

File: examples/gpt/schedule.py [hunk 5]
```diff
@@ -250,7 +252,7 @@ def fwd_post_hook(_module, _input, output):
         input_mask = (_input[0] < vocab_start_index) | (_input[0] >= vocab_end_index)
         output[input_mask, :] = 0.0
         # Reduce across all the model parallel GPUs
-        dist.all_reduce(output, op=dist.ReduceOp.SUM, group=sch.group)
+        output = slapo.sharding.reduce_forward_output(output, sch.group)
```
[the remark this answers is not in the document]
Comment (szhengac) [reply examples/gpt/schedule.py:255]: it handles the backward part, otherwise the grad will not flow through

[the remark this answers is not in the document]
Comment (comaniac) [reply examples/gpt/schedule.py:255]: Oh I got it. This is a bit tricky tho. We need to mention in doc or tutorial that if anyone wants to have comm ops in their hooks, they have to use ours...

**Comment by comaniac** [comment]
Thanks @szhengac @chhzh123 @zarzen
